binary_transparency/firmware/internal/client/follower.go

```diff
@@ -0,0 +1,185 @@
+	return fmt.Sprintf("failed to verify inclusion proof %s in root %s", e.Proof, e.Checkpoint)
+}
+
+// LogEntry wraps up a leaf value with its position in the log.
+type LogEntry struct {
+	Root  api.LogCheckpoint
+	Index uint64
+	Value api.SignedStatement
+}
+
+// LogFollower follows a log for new data becoming available.
+type LogFollower struct {
+	c            ReadonlyClient
+	lv           logverifier.LogVerifier
+	pollInterval time.Duration
+	golden       api.LogCheckpoint
+}
+
+func NewLogFollower(c ReadonlyClient, pollInterval time.Duration, golden api.LogCheckpoint) LogFollower {
+	return LogFollower{
+		c:            c,
+		lv:           verify.NewLogVerifier(),
+		pollInterval: pollInterval,
+		golden:       golden,
+	}
+}
+
+// Checkpoints polls the log according to the configured interval, returning roots consistent
+// with the current golden checkpoint. Should any valid roots be found which are inconsistent
+// then an error is returned. The log being unavailable will just cause a retry, giving it
+// the benefit of the doubt.
+func (f *LogFollower) Checkpoints(ctx context.Context) (<-chan api.LogCheckpoint, <-chan error) {
+	ticker := time.NewTicker(f.pollInterval)
+
+	outc := make(chan api.LogCheckpoint, 1)
+	errc := make(chan error, 1)
+
+	// Push the starting checkpoint into the channel here to kick off the pipeline
+	outc <- f.golden
+
+	go func() {
+		defer close(outc)
+		// Now keep looking for newer, consistent checkpoints
+		for {
+			select {
+			case <-ticker.C:
+				// Wait until the next tick.
+			case <-ctx.Done():
+				errc <- ctx.Err()
+				return
+			}
+
+			cp, err := f.c.GetCheckpoint()
+			if err != nil {
+				glog.Warningf("Failed to get latest Checkpoint: %q", err)
+				continue
+			}
+			// TODO(al): check signature on checkpoint when they're added.
+
+			if cp.TreeSize <= f.golden.TreeSize {
+				continue
+			}
+			glog.V(1).Infof("Got newer checkpoint %s", cp)
+
+			// Perform consistency check only for non-zero initial tree size
+			if f.golden.TreeSize != 0 {
+				consistency, err := f.c.GetConsistencyProof(api.GetConsistencyRequest{From: f.golden.TreeSize, To: cp.TreeSize})
+				if err != nil {
+					glog.Warningf("Failed to fetch the Consistency: %q", err)
+					continue
+				}
+				glog.V(1).Infof("Printing the latest Consistency Proof Information")
+				glog.V(1).Infof("Consistency Proof = %x", consistency.Proof)
+
+				// Verify the fetched consistency proof
+				if err := f.lv.VerifyConsistencyProof(int64(f.golden.TreeSize), int64(cp.TreeSize), f.golden.RootHash, cp.RootHash, consistency.Proof); err != nil {
+					errc <- ErrConsistency{
+						Golden: f.golden,
+						Latest: *cp,
+					}
+					return
+				}
+				glog.V(1).Infof("Consistency proof for Treesize %d verified", cp.TreeSize)
+			}
+			f.golden = *cp
+			outc <- *cp
+		}
+	}()
+	return outc, errc
+}
+
+func (f *LogFollower) Entries(ctx context.Context, cpc <-chan api.LogCheckpoint, head uint64) (<-chan LogEntry, <-chan error) {
```

Comment:
Comment on Entries, similar to Checkpoints will be helpful!

Reply:
Yup I was just working on that actually :-) Now done.